initial commit for CSI adapter

krishchow opened the pull request:
This PR aims to add the basic boilerplate for the COSI CSI adapter. Future PRs will add implementations of the Node controller and Identity controller for the CSI adapter.

NicolasT commented:
Your branch contains `bin/csi-adapater`, a 47MB binary file, which should not be in VCS. You'll likely want to remove it, and squash such removal with the code commit so the binary doesn't end up in the repository.

NicolasT commented on pkg/controller/controller.go:
```diff
@@ -0,0 +1,90 @@
+}
+
+func (c *ControllerServer) ControllerGetCapabilities(ctx context.Context, req *csi.ControllerGetCapabilitiesRequest) (*csi.ControllerGetCapabilitiesResponse, error) {
+	return nil, status.Error(codes.Unimplemented, "unimplemented")
```
Instead of all of these, you could make `type ControllerServer` 'inherit' from `csi.UnimplementedControllerServer`.

krishchow replied:
I'll look into this for the CSI driver! Thank for sharing

NicolasT commented on pkg/identity/identity.go:
```diff
@@ -0,0 +1,49 @@
+}
+
+func (i *IdentityServer) GetPluginInfo(ctx context.Context, req *csi.GetPluginInfoRequest) (*csi.GetPluginInfoResponse, error) {
+	return &csi.GetPluginInfoResponse{}, nil
```
Instead of stub results, consider `IdentityServer` to 'inherit' from `csi.UnimplementedIdentityServer`.

krishchow replied:
For the `IdentityServer`, I would actually prefer to do it this way. Most of the logic is already outlined in another commit, and when I open the second PR to add that here, we might as well try to keep the diff as small as possible.

NicolasT commented on pkg/node/node.go:
```diff
@@ -0,0 +1,80 @@
+}
+
+func (n NodeServer) NodeStageVolume(ctx context.Context, request *csi.NodeStageVolumeRequest) (*csi.NodeStageVolumeResponse, error) {
+	panic("implement me")
```
Same, let `NodeServer` 'inherit' `csi.UnimplementedNodeServer`.

krishchow replied:
Same as the `IdentityServer`, the actual implementation for the NodeServer is mostly already finished in an external PR, and I think it would be more valuable to keep future diffs smaller

krishchow commented:
> Your branch contains `bin/csi-adapater`, a 47MB binary file, which should not be in VCS. You'll likely want to remove it, and squash such removal with the code commit so the binary doesn't end up in the repository.

Rebased and removed, I thought I added that to the .gitignore :sweat_smile:

NicolasT commented on main.go:
```diff
@@ -0,0 +1,41 @@
+
+func main() {
+	sigs := make(chan os.Signal, 1)
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM, syscall.SIGSEGV)
```
Not sure we want to 'properly handle' `SIGSEGV`. IMO a segmentation fault should crash an application unless the app can really do something 'intelligent' about it.

wlan0 replied:
SEGFault should not lead to forced termination of any I/O in progress. The whole point of gracefully handling this signal to complete I/O in progress before exiting. This should still be here and added to the other repos as well if it’s not already there

krishchow replied:
That makes sense to me, I have removed that signal for the time being

NicolasT commented on main.go:
```diff
@@ -0,0 +1,41 @@
+	go func() {
+		s := <-sigs
+		glog.Infof("Exiting on signal %s %#v", s.String(), s)
+		panic("Signal received. Exiting")
```
No need to `panic`.

wlan0 replied:
Yeah instead context should be cancelled

krishchow replied:
Removed the panic

NicolasT commented on cmd/driver.go:
```diff
@@ -0,0 +1,62 @@
+	if err != nil {
+		return err
+	}
+	glog.V(5).Infof("identity server started")
```
It's not really `started` at this point.

krishchow replied:
Reworded to "prepared"

NicolasT commented on .gitignore:
```diff
@@ -0,0 +1,3 @@
+.idea
+vendor
+bin/csi-adapater
```
```suggestion
bin/csi-adapter
```

NicolasT commented on Makefile:
```diff
@@ -0,0 +1,2 @@
+build:
+	go build -o bin/csi-adapater main.go
```
```suggestion
	go build -o bin/csi-adapter main.go
```

tparikh commented on main.go:
```diff
@@ -0,0 +1,40 @@
+/*
```
@krishchow To build the container image using release tools, this main.go needs to be moved to `cmd/csi-adapater/main.go`

krishchow replied:
Does anything else need to change (aside from updating references) in the main.go file for the release tools?

Commit:
initial commit, post review fixes and release tools

wlan0 commented:
All the outstanding comments have been addressed. LGTM

wlan0 commented:
/lgtm
